### PR TITLE
Jokerace/BKRT integration

### PR DESCRIPTION
**Pull request by vivinvinh212**

### CODE REVIEW

File: src/SwanDebate.sol [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    function _determineWinner(
+        IJokeRaceContest _contest,
+        uint256[] memory _proposalIds,
+        address _agent1,
+        address _agent2
+    ) internal view returns (address winner, uint256 highestVotes) {
+        for (uint256 i = 0; i < _proposalIds.length; i++) {
+            (uint256 forVotes,) = _contest.proposalVotes(_proposalIds[i]);
+            if (forVotes > highestVotes) {
+                (address author,,) = _contest.proposals(_proposalIds[i]);
```
Comment (erhant) [inline src/SwanDebate.sol:265]: need some comments on the JokeRace interface

File: src/SwanDebate.sol [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+    /// @dev Only owner can record outputs and both agents must provide output to complete a round
+    function recordOracleOutput(address _contest, uint256 _agentId, uint256 _taskId, bytes calldata _output)
+        external
+        onlyOwner
+        whenNotPaused
+    {
+        if (!agents[_agentId].isRegistered) revert AgentNotRegistered();
+
+        Debate storage debate = debates[_contest];
+        if (debate.currentRound == 0 || debate.winnerId != 0) revert DebateNotActive(_contest);
```
Comment (erhant) [inline src/SwanDebate.sol:277]: we can write this with an internal function `isDebateActive` perhaps, to make it clear in its documentation that an active debate has a non-zero round & is without a winenr

File: src/SwanDebate.sol [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+        if (_agentId != debate.agent1Id && _agentId != debate.agent2Id) revert InvalidAgent(_agentId);
+
+        IJokeRaceContest contest = IJokeRaceContest(_contest);
+        if (contest.state() != IJokeRaceContest.ContestState.Active) {
+            revert ContestInvalidState(contest.state(), IJokeRaceContest.ContestState.Active);
+        }
+
+        // Combine system prompt with input
+        bytes memory combinedInput = abi.encodePacked(agents[_agentId].systemPrompt, "\n\n", _input);
```
Comment (erhant) [inline src/SwanDebate.sol:256]: lets confirm this `\n\n` with Batu as well, to see if there is a better way to connect the system prompt to `input`

File: src/SwanDebate.sol [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+        onlyOwner
+        whenNotPaused
+    {
+        if (!agents[_agentId].isRegistered) revert AgentNotRegistered();
+
+        Debate storage debate = debates[_contest];
+        if (debate.currentRound == 0 || debate.winnerId != 0) revert DebateNotActive(_contest);
+
+        IJokeRaceContest contest = IJokeRaceContest(_contest);
+        if (contest.state() != IJokeRaceContest.ContestState.Active) {
```
Comment (erhant) [inline src/SwanDebate.sol:280]: we can also write a modifier for this perhaps

File: src/SwanDebate.sol [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+    event DebateInitialized(address indexed contest, uint256 indexed agent1Id, uint256 indexed agent2Id);
+
+    /// @notice Emitted when an oracle output is recorded for an agent
+    /// @param contest Address of the JokeRace contest
+    /// @param round Current round number
+    /// @param agentId ID of agent providing output
+    /// @param taskId Oracle task identifier
+    /// @param output Raw output data from oracle
+    event OracleOutputRecorded(
+        address indexed contest, uint256 indexed round, uint256 indexed agentId, uint256 taskId, bytes output
```
Comment (erhant) [inline src/SwanDebate.sol:100]: same comment here, emitting `bytes` will likely emit its 32-byte hash instead we can omit it probably

File: src/SwanDebate.sol [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+        int256 netVotes2 = int256(forVotes2) - int256(againstVotes2);
+
+        if (netVotes1 > netVotes2) {
+            return (debate.agent1Id, uint256(netVotes1));
+        } else if (netVotes2 > netVotes1) {
+            return (debate.agent2Id, uint256(netVotes2));
+        } else {
+            // Implement a tie-breaking rule
+            winnerId = debate.agent1Id;
+            highestVotes = uint256(netVotes1);
```
Comment (erhant) [inline src/SwanDebate.sol:422]: this is TODO I think right?

**Review by erhant — CHANGES_REQUESTED** [review]
✨ see comments ✨

File: src/SwanDebate.sol (hunk 3 above)
Comment (erhant) [reply src/SwanDebate.sol:256]: @aktasbatuhan

File: src/SwanDebate.sol [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                            INTERNAL FUNCTIONS
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Determines the winner of a debate based on JokeRace voting results
+    /// @param _contest JokeRace contest interface
+    /// @param debate Debate to determine winner
+    function _determineWinner(IJokeRaceContest _contest, Debate storage debate)
```
[the remark this answers is not in the document]
Comment (erhant) [reply src/SwanDebate.sol:425]: we have decided that this logic is fine, and this function is only called when a debate is finished (which is expected to happen when an end-date is reached for that contest)

File: test/SwanDebateTest.t.sol [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+            contests[i] = address(newContest);
+
+            debate.initializeDebate(agent1Id, agent2Id, address(newContest));
+        }
+
+        address[] memory agent1Debates = debate.getAgentDebates(agent1Id);
+        assertEq(agent1Debates.length, 3, "Should track all debates");
+    }
+
+    function test_Events() external {
```
Comment (erhant) [inline test/SwanDebateTest.t.sol:287]: is it possible that we use [vm.expectEmit](https://book.getfoundry.sh/cheatcodes/expect-emit) here?

File: test/SwanDebateTest.t.sol [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+        // Start second round
+        setupOracleOutput(contestAddr, agent1Id, input, models, oracleParameters);
+        setupOracleOutput(contestAddr, agent2Id, input, models, oracleParameters);
+
+        // Verify second round completion
+        SwanDebate.RoundData memory round2Data = debate.getRoundForDebate(contestAddr, 2);
+        assertTrue(round2Data.roundComplete, "Second round should be complete");
+    }
+
+    function test_EdgeCases() external fund addValidatorsToWhitelist registerOracles {
```
Comment (erhant) [inline test/SwanDebateTest.t.sol:383]: lets write these as separate tests; while its indeed a minor thing its better if we have a `test_***` for each edge case, so that in-case one of them fails later we see which better

File: src/SwanDebate.sol [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+    /// @param _contest Address of the JokeRace contest
+    /// @return The address of the initialized contest
+    /// @dev Verifies agent registration and contest state before initialization
+    function initializeDebate(uint256 _agent1Id, uint256 _agent2Id, address _contest)
+        external
+        onlyOwner
+        whenNotPaused
+        returns (address)
+    {
+        if (!agents[_agent1Id].isRegistered || !agents[_agent2Id].isRegistered) {
```
Comment (erhant) [inline src/SwanDebate.sol:228]: lets also add a modifier for `onlyRegisteredAgent(address)`
this applies to other functions too, but here we will need to call the same modifier two times

File: test/SwanDebateTest.t.sol [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+import {SwanDebate} from "../src/SwanDebate.sol";
+import {LLMOracleCoordinator} from "@firstbatch/dria-oracle-contracts/LLMOracleCoordinator.sol";
+import {SwanAgent} from "../src/SwanAgent.sol";
+import {IJokeRaceContest} from "../src/SwanDebate.sol";
+import {LLMOracleTaskParameters} from "@firstbatch/dria-oracle-contracts/LLMOracleTask.sol";
+import {Vm} from "forge-std/Vm.sol";
+import {console} from "forge-std/Test.sol";
+
+/// @dev Mock implementation of `IJokeRaceContest`
+contract MockJokeRaceContest is IJokeRaceContest {
```
Comment (erhant) [inline test/SwanDebateTest.t.sol:15]: lets move this to a different file in `./test/mock/JokeRaceContest.sol`

File: test/SwanDebateTest.t.sol [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+            if (_proposalIds[i] == proposalId) {
+                return true;
+            }
+        }
+        return false;
+    }
+}
+
+/// @dev Mock Oracle Contract
+contract MockOracle {
```
Comment (erhant) [inline test/SwanDebateTest.t.sol:76]: lets move this to a different file `./test/mock/LLMOracleCoordinator.sol`

**Review by erhant — CHANGES_REQUESTED** [review]
nice work! see comments below 🙏🏻
